packages/docs/docs/docs/proxies.md

```diff
@@ -0,0 +1,68 @@
+
+Note: For a more in depth read, please see [blog.zeppelinos.org/proxy-patterns](https://blog.zeppelinos.org/proxy-patterns/), which discusses the need for proxies, goes into more technical detail on the subject, elaborates on other possible proxy patterns that were considered for zOS, and more.
+
+## Why upgradeability?
+
+By design, smart contracts are immutable. On the other hand, software quality heavily depends on the ability to upgrade and patch source code in order to make iterative releases. Even though blockchain based software profits signifficantly from the technology's immutability, still a certain degree of mutability is needed for bugfixing and potential product improvements. ZeppelinOS solves this apparent contradiction by providing an easy to use, simple and robust upgrade mechanism for smart contracts that can be controlled by any type of governance, be it a multi-sig wallet, a simple address or a complex DAO.
+
+## Upgradeability via the proxy pattern
+
+The basic idea behind using the proxy pattern to solve the problem of pugradeability, is to use two contracts instead of one. The first contract is a simple wrapper or "proxy" which users interact with directly and is in charge of forwarding transactions to and from the second contract, which contains the logic of the pair. The key concept to understand is that the logic contract can be replaced while the proxy is never changed. Both contracts are still immutable, in the sense that their code cannot be changed, but the logic contract can simply be swapped by another. The wrapper can thus point to a different logic implementation and in doing so, the software is "upgraded".
+
+## Proxy forwarding
+
+The most immediate problem that proxies need to solve is how the proxy exposes the entire interface of the logic contract without requiring a one to one mapping of the entire logic contract's interface. This would be difficult to maintain, prone to errors, and would make the interface itself not upgradeable. Hence, a dynamic forwarding mechanism is required. The basics of such mechanism can be summarized in the code below:
+
+```
+assembly {
+    let ptr := mload(0x40)
+    calldatacopy(ptr, 0, calldatasize)
+    let result := delegatecall(gas, _impl, ptr, calldatasize, 0, 0)
+    let size := returndatasize
+    returndatacopy(ptr, 0, size)
+
+    switch result
+    case 0 { revert(ptr, size) }
+    default { return(ptr, size) }
+ }
+ ```
+
+This code can be put in the fallback function of a proxy, and will forward any call to any function with any set of parameters to the logic contract without it needing to know anything in particular of the logic contract's interface. In essence, (1) the `calldata` is copied to memory, (2) the call is forwarded to the logic contract, (3) the return data from the call to the logic contract is retrieved, and (4) the returned data is forwarded back to the caller.
```

Comment:
link to the solidity docs of fallback function.
It would be good to explain why this is assembly and not solidity code.